### DISCUSSION TITLE
Are there any events on new external address added.

### DISCUSSION THREAD

**Discussion by lynic** [libp2p/rust-libp2p#3030] [Q&A]
I notice that the peer's external addresses could be auto-updated, I wondering which part is doing it.
And some of these external address are incorrect, I want to manually remove them. If there's an event on new external address added, then I could use `swarm.remove_external_address` to remove it on external address events.

**Comment by thomaseizinger** [comment]
At the moment, I think the only protocol that adds extenal addresses is the identify protocol so you could couple your logic to that.

It makes me wonder though, in what way are they wrong?

**Reply by lynic** [reply]
I try to follow the examples to setup relay server and dcutr clients. I enable identify on both relay server and client. I started them in docker containers.
When client gets a identify events from relay server, it will start listening on a p2p-circuit address like `/dns/<relay-server>/tcp/9090/p2p/xxx/p2p-circuit/p2p/<self-id>`
But seems swarm will automatically listen on other private address on relay server, e.g. `/ip4/172.17.0.1/tcp/9090/p2p/12D3KooWGjfbZFdzpfURhCAtGYqk62WmmKxHNo1EEhQiUisij7DY/p2p-circuit/p2p/12D3KooWSJQeQTDsbWEQwAcRWwMGp6bqAJziDmXoeerva7bbes5q`
It results identify report this addr to others and added to external addresss? It will increase the identify datas send to other peers, so I want to  use `swam.remoev_external_address` to remove them dynamically base on events.

**Reply by thomaseizinger** [reply]
If I understand you correctly, then the issue is that the relay server is listening on addresses that it is not actually reachable on. You should be able to fix your problem by not listening on private interfaces on the relay server.

**Reply by lynic** [reply]
Thanks @thomaseizinger ! Relay server running inside a container, hence it only sees private address, will consider make it running on server directly.

I go through identify code, but couldn't find the code to update external addresses, could you help to point me to the related code?

**Reply by thomaseizinger** [reply]
> Thanks @thomaseizinger ! Relay server running inside a container, hence it only sees private address, will consider make it running on server directly.

The container bit doesn't matter really. Are you listening using a wildcard address and thus on all interfaces?

> I go through identify code, but couldn't find the code to update external addresses, could you help to point me to the related code?

The external address is registered here: https://github.com/libp2p/rust-libp2p/blob/4d4833f8c2622018b5b775896905a534b8624d3c/protocols/identify/src/behaviour.rs#L319

But you don't need to modify that code.

I suggest you fix your deployment of the relay server to only listen on meaningful interfaces.
Alternatively you can remove external addresses in the event-loop where you poll the swarm, whenever you get an identify event.

**Reply by lynic** [reply]
Yes, my server listen on 0.0.0.0 address. I will try to make it only listen on the external address.
Thanks pointing me to the code in identify!
